File: src/main/java/com/pablogeorge/workout_app/service/WorkoutService.java
package com.pablogeorge.workout_app.service;

import com.pablogeorge.workout_app.model.CoreWorkout;
import com.pablogeorge.workout_app.model.TrainingMaxLog;
import com.pablogeorge.workout_app.model.UserLiftConfig;
import com.pablogeorge.workout_app.model.WorkoutLog;
import com.pablogeorge.workout_app.repository.TrainingMaxLogRepository;
import com.pablogeorge.workout_app.repository.UserLiftConfigRepository;
import com.pablogeorge.workout_app.repository.WorkoutLogRepository;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class WorkoutService {

    private final UserLiftConfigRepository repo;
    private final WorkoutLogRepository logRepo;
    private final TrainingMaxLogRepository tmLogRepo;

    public WorkoutService(UserLiftConfigRepository repo,
                          WorkoutLogRepository logRepo,
                          TrainingMaxLogRepository tmLogRepo) {
        this.repo = repo;
        this.logRepo = logRepo;
        this.tmLogRepo = tmLogRepo;
    }

    // ── Program definition ────────────────────────────────────────────────────

    private record SetDef(double pct, int reps, boolean amrap, boolean warmup) {}

    private static final List<List<SetDef>> PROGRAM = List.of(
        List.of(
            new SetDef(0.50, 5, false, true),
            new SetDef(0.50, 5, false, true),
            new SetDef(0.65, 5, false, false),
            new SetDef(0.75, 5, false, false),
            new SetDef(0.85, 0, true,  false)
        ),
        List.of(
            new SetDef(0.50, 5, false, true),
            new SetDef(0.50, 5, false, true),
            new SetDef(0.70, 5, false, false),
            new SetDef(0.80, 3, false, false),
            new SetDef(0.90, 0, true,  false)
        ),
        List.of(
            new SetDef(0.50, 5, false, true),
            new SetDef(0.50, 5, false, true),
            new SetDef(0.75, 3, false, false),
            new SetDef(0.85, 3, false, false),
            new SetDef(0.95, 0, true,  false)
        ),
        List.of(
            new SetDef(0.40, 5, false, true),
            new SetDef(0.40, 5, false, true),
            new SetDef(0.50, 5, false, false),
            new SetDef(0.60, 5, false, false),
            new SetDef(0.70, 0, true,  false)
        )
    );

    // ── Public records ────────────────────────────────────────────────────────

    public record WorkoutSet(
        boolean warmup,
        String percentageLabel,
        double weight,
        int reps,
        boolean amrap,
        String platesDisplay
    ) {
        public String repsLabel() {
            return amrap ? "AMRAP" : reps + " reps";
        }
    }

    public record WorkoutPlan(
        String liftName,
        int week,
        double trainingMax,
        List<WorkoutSet> sets
    ) {
        public String weekLabel() {
            return switch (week) {
                case 1 -> "5s Week";
                case 2 -> "3s Week";
                case 3 -> "5/3/1 Week";
                case 4 -> "Deload";
                default -> "Week " + week;
            };
        }
    }

    // ── Config management ─────────────────────────────────────────────────────

    public Optional<UserLiftConfig> getConfig(String userId, Long liftId) {
        return repo.findByUserIdAndLift_Id(userId, liftId);
    }

    public UserLiftConfig createConfig(String userId, CoreWorkout lift, double trainingMax) {
        UserLiftConfig config = repo.save(new UserLiftConfig(userId, lift, trainingMax));
        tmLogRepo.save(new TrainingMaxLog(userId, lift, trainingMax));
        return config;
    }

    public void updateTrainingMax(UserLiftConfig config, double newMax) {
        config.setTrainingMax(newMax);
        repo.save(config);
        tmLogRepo.save(new TrainingMaxLog(config.getUserId(), config.getLift(), newMax));
    }

    public Map<Long, String> getWeekLabels(String userId) {
        return repo.findByUserId(userId).stream()
            .collect(Collectors.toMap(
                c -> c.getLift().getId(),
                c -> "Week " + c.getCurrentWeek() + " · " + weekLabel(c.getCurrentWeek())
            ));
    }

    // ── Workout execution ─────────────────────────────────────────────────────

    public WorkoutPlan buildPlan(UserLiftConfig config, CoreWorkout lift) {
        List<SetDef> defs = PROGRAM.get(config.getCurrentWeek() - 1);
        List<WorkoutSet> sets = defs.stream().map(def -> {
            double weight = roundUpTo5(config.getTrainingMax() * def.pct());
            return new WorkoutSet(
                def.warmup(),
                (int)(def.pct() * 100) + "%",
                weight,
                def.reps(),
                def.amrap(),
                platesDisplay(weight)
            );
        }).toList();
        return new WorkoutPlan(lift.getName(), config.getCurrentWeek(), config.getTrainingMax(), sets);
    }

    public void completeWorkout(UserLiftConfig config, int amrapReps) {
        logRepo.save(new WorkoutLog(config.getUserId(), config.getLift(), config.getCurrentWeek(), amrapReps));
        int next = (config.getCurrentWeek() % 4) + 1;
        config.setCurrentWeek(next);
        repo.save(config);
    }

    public long countLogs(String userId) {
        return logRepo.countByUserId(userId);
    }

    // ── Chart: training max over time ─────────────────────────────────────────

    private static final String[] CHART_COLORS = {
        "#4f9eff", "#4ade80", "#fb923c", "#a78bfa", "#f472b6"
    };

    public List<Map<String, Object>> buildChartDatasets(String userId) {
        List<TrainingMaxLog> logs = tmLogRepo.findByUserIdOrderByLoggedOnAsc(userId);

        Map<String, List<TrainingMaxLog>> byLift = new LinkedHashMap<>();
        for (TrainingMaxLog log : logs) {
            byLift.computeIfAbsent(log.getLift().getName(), k -> new ArrayList<>()).add(log);
        }

        List<Map<String, Object>> datasets = new ArrayList<>();
        int i = 0;
        for (var entry : byLift.entrySet()) {
            List<Map<String, Object>> points = entry.getValue().stream()
                .map(log -> Map.<String, Object>of(
                    "x", log.getLoggedOn().toString(),
                    "y", log.getTrainingMax()
                ))
                .toList();

            datasets.add(Map.of(
                "label", entry.getKey(),
                "color", CHART_COLORS[i % CHART_COLORS.length],
                "points", points
            ));
            i++;
        }
        return datasets;
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private String weekLabel(int week) {
        return switch (week) {
            case 1 -> "5s";
            case 2 -> "3s";
            case 3 -> "5/3/1";
            case 4 -> "Deload";
            default -> "Week " + week;
        };
    }

    private double roundUpTo5(double weight) {
        return Math.ceil(weight / 5.0) * 5.0;
    }

    private String platesDisplay(double totalWeight) {
        double perSide = (totalWeight - 45) / 2;
        double[] plateTypes = {45, 35, 25, 10, 5, 2.5};
        List<String> plates = new ArrayList<>();
        double remaining = perSide;

        for (double plate : plateTypes) {
            while (remaining >= plate - 0.01) {
                plates.add(plate == (int) plate
                    ? String.valueOf((int) plate)
                    : String.valueOf(plate));
                remaining -= plate;
            }
        }
        return plates.isEmpty() ? "bar only" : String.join(", ", plates);
    }
}

File: src/services/workoutService.ts
import { prisma } from "../app.js";

interface SetDef {
  pct: number;
  reps: number;
  amrap: boolean;
  warmup: boolean;
}

const PROGRAM: SetDef[][] = [
  [
    { pct: 0.50, reps: 5, amrap: false, warmup: true },
    { pct: 0.50, reps: 5, amrap: false, warmup: true },
    { pct: 0.65, reps: 5, amrap: false, warmup: false },
    { pct: 0.75, reps: 5, amrap: false, warmup: false },
    { pct: 0.85, reps: 0, amrap: true, warmup: false },
  ],
  [
    { pct: 0.50, reps: 5, amrap: false, warmup: true },
    { pct: 0.50, reps: 5, amrap: false, warmup: true },
    { pct: 0.70, reps: 5, amrap: false, warmup: false },
    { pct: 0.80, reps: 3, amrap: false, warmup: false },
    { pct: 0.90, reps: 0, amrap: true, warmup: false },
  ],
  [
    { pct: 0.50, reps: 5, amrap: false, warmup: true },
    { pct: 0.50, reps: 5, amrap: false, warmup: true },
    { pct: 0.75, reps: 3, amrap: false, warmup: false },
    { pct: 0.85, reps: 3, amrap: false, warmup: false },
    { pct: 0.95, reps: 0, amrap: true, warmup: false },
  ],
  [
    { pct: 0.40, reps: 5, amrap: false, warmup: true },
    { pct: 0.40, reps: 5, amrap: false, warmup: true },
    { pct: 0.50, reps: 5, amrap: false, warmup: false },
    { pct: 0.60, reps: 5, amrap: false, warmup: false },
    { pct: 0.70, reps: 0, amrap: true, warmup: false },
  ],
];

export interface WorkoutSet {
  warmup: boolean;
  percentageLabel: string;
  weight: number;
  reps: number;
  amrap: boolean;
  platesDisplay: string;
  repsLabel: string;
}

export interface WorkoutPlan {
  liftName: string;
  week: number;
  trainingMax: number;
  sets: WorkoutSet[];
  weekLabel: string;
}

function roundUpTo5(weight: number): number {
  return Math.ceil(weight / 5) * 5;
}

function platesDisplay(totalWeight: number): string {
  const perSide = (totalWeight - 45) / 2;
  const plateTypes = [45, 35, 25, 10, 5, 2.5];
  const plates: string[] = [];
  let remaining = perSide;

  for (const plate of plateTypes) {
    while (remaining >= plate - 0.01) {
      plates.push(Number.isInteger(plate) ? String(plate) : String(plate));
      remaining -= plate;
    }
  }

  return plates.length === 0 ? "bar only" : plates.join(", ");
}

function weekLabel(week: number): string {
  switch (week) {
    case 1: return "5s Week";
    case 2: return "3s Week";
    case 3: return "5/3/1 Week";
    case 4: return "Deload";
    default: return `Week ${week}`;
  }
}

export async function getConfig(userId: string, liftId: number) {
  return prisma.userLiftConfig.findUnique({
    where: { userId_liftId: { userId, liftId } },
  });
}

export async function createConfig(userId: string, lift: { id: number; name: string }, trainingMax: number) {
  return prisma.userLiftConfig.create({
    data: { userId, liftId: lift.id, trainingMax, currentWeek: 1 },
  });
}

export async function updateTrainingMax(
  config: { id: number; userId: string; liftId: number },
  newMax: number
) {
  await prisma.userLiftConfig.update({
    where: { id: config.id },
    data: { trainingMax: newMax },
  });
}

export async function getWeekLabels(userId: string) {
  const configs = await prisma.userLiftConfig.findMany({
    where: { userId },
  });

  const labels: Record<number, string> = {};
  for (const c of configs) {
    labels[c.liftId] = `Week ${c.currentWeek} · ${weekLabel(c.currentWeek)}`;
  }
  return labels;
}

export async function buildPlan(
  config: { currentWeek: number; trainingMax: number; liftId: number },
  lift: { name: string }
): Promise<WorkoutPlan> {
  const defs = PROGRAM[config.currentWeek - 1];
  const sets: WorkoutSet[] = defs.map((def) => {
    const weight = roundUpTo5(config.trainingMax * def.pct);
    return {
      warmup: def.warmup,
      percentageLabel: `${Math.round(def.pct * 100)}%`,
      weight,
      reps: def.reps,
      amrap: def.amrap,
      platesDisplay: platesDisplay(weight),
      repsLabel: def.amrap ? "AMRAP" : `${def.reps} reps`,
    };
  });

  return {
    liftName: lift.name,
    week: config.currentWeek,
    trainingMax: config.trainingMax,
    sets,
    weekLabel: weekLabel(config.currentWeek),
  };
}

export async function completeWorkout(
  config: { id: number; userId: string; liftId: number; currentWeek: number },
  amrapReps: number
) {
  await prisma.workoutLog.create({
    data: {
      userId: config.userId,
      liftId: config.liftId,
      week: config.currentWeek,
      amrapReps,
      completedOn: new Date().toISOString().split("T")[0],
    },
  });

  const next = (config.currentWeek % 4) + 1;
  await prisma.userLiftConfig.update({
    where: { id: config.id },
    data: { currentWeek: next },
  });
}

export async function countLogs(userId: string) {
  return prisma.workoutLog.count({ where: { userId } });
}

const CHART_COLORS = [
  "#4f9eff", "#4ade80", "#fb923c", "#a78bfa", "#f472b6",
];

export async function buildChartDatasets(userId: string) {
  const logs = await prisma.trainingMaxLog.findMany({
    where: { userId },
    orderBy: { loggedOn: "asc" },
    include: { lift: true },
  });

  const byLift = new Map<string, typeof logs>();
  for (const log of logs) {
    const name = log.lift.name;
    if (!byLift.has(name)) byLift.set(name, []);
    byLift.get(name)!.push(log);
  }

  const datasets: { label: string; color: string; points: { x: string; y: number }[] }[] = [];
  let i = 0;
  for (const [label, points] of byLift) {
    datasets.push({
      label,
      color: CHART_COLORS[i % CHART_COLORS.length],
      points: points.map((p) => ({ x: p.loggedOn, y: p.trainingMax })),
    });
    i++;
  }

  return datasets;
}
